### ISSUE TITLE
Not loading shared notes in multiple instances.

### ISSUE THREAD

**Issue by johnkolen** [paulrosen/abcjs#721]
Hi,
I've got a situation where I have multiple, very short, note sequences that the user can play and evaluate. The longest sequences are up and down scales, supplemented with short subsequences. When I create a C Major scale in one div and the first four notes of the same scale in a different div (C2D2E2F2), both render properly, animate properly, but the four note sequence fails to play. If I change the short sequence by dropping it an octave (C,2D,2E,2F,2), everything works as expected.

The problem occurs in both 6.0.0-beta.33 and 6.0.0-beta.34(dev).

The console log reports:
```
Didn't load note acoustic_grand_piano C4 pending
Didn't load note acoustic_grand_piano D4 pending
Didn't load note acoustic_grand_piano E4 pending
Didn't load note acoustic_grand_piano F4 pending
```
The synth.prime promise also succeeds. Not sure if this is a bug, limitation, or misconfiguration. 
Any help would be appreciated.
Thanks in advance,
John

**Comment by johnkolen** [comment]
On a lark, I delayed the creation of the shorter sequence by about 3 seconds and it worked. This situation is more in line with the desired functionality as the shorter sequences are generated on the fly. 

However, that leaves  us in the position that the notes fail to load but the synth reports that it's good to go.

**Comment by hajimecm** [comment]
After a bit of manual bisecting among tags, it appears that v6.0.0-beta.19 is the last tag where this was working.

**Comment by paulrosen** [comment]
Thanks for the hint on this. I'll see if I can recreate. Your instructions above look clear enough but if you have a simple page that shows the problem that would help.

**Comment by foxbunny** [comment]
@paulrosen  Ok, I think I've found a way to fix this problem. I'm not a 100% sure if this is the appropriate way to fix it, though, so I didn't do a PR yet, but I'm willing to if you think this approach is ok.

So I'm still not quite certain, but this appears to be a race where mutliple instances that are independently primed will cause the assumptions `prime()` makes about the sound cache to become untrue. To work around this, I made two changes (please excuse the accidental modernization of the code base, I didn't want to spend time doing it the old-school way):

```
// load-note.js

var getNote = function getNote(url, instrument, name, audioContext) {
  if (!soundsCache[instrument]) soundsCache[instrument] = {};
  var instrumentCache = soundsCache[instrument];

  if (instrumentCache[name]) return instrumentCache[name]
  return instrumentCache[name] = fetch(url + instrument + '-mp3/' + name + '.mp3')
    .then(resp => resp.arrayBuffer())
    .then(buff => new Promise((res, rej) => audioContext.decodeAudioData(buff, res, rej)))
};


// place-note.js

function placeNote(outputAudioBuffer, sampleRate, sound, startArray, volumeMultiplier, ofsMs, fadeTimeSec, noteEndSec) {
  // sound contains { instrument, pitch, volume, len, pan, tempoMultiplier
  // len is in whole notes. Multiply by tempoMultiplier to get seconds.
  // ofsMs is an offset to subtract from the note to line up programs that have different length onsets.
  var OfflineAC = window.OfflineAudioContext || window.webkitOfflineAudioContext;
  var len = sound.len * sound.tempoMultiplier;
  if (ofsMs) len += ofsMs / 1000;
  len -= noteEndSec;
  if (len < 0) len = 0.005; // Have some small audible length no matter how short the note is.

  var offlineCtx = new OfflineAC(2, Math.floor((len + fadeTimeSec) * sampleRate), sampleRate);
  var noteName = pitchToNoteName[sound.pitch];
  return soundsCache[sound.instrument][noteName]
    .then(noteBuffer => { /* branch where note is available */ })
    .catch(error => console.trace("Didn't load note", sound.instrument, noteName, error));
```

In short, in `getNote()` I stored the `Promise` object rather than an `AudioBuffer` instance, and got rid of the `pending` and `error` strings. Since once a `Promise` reference is stored, it can be `then()`'d and `catch()`'d any number of times, we can do this in `placeNote()` instead of trying to syncrhonously inspect the state. In my current application, this appears to work as expected. 

Let me know if this appears to be a good fix, and I'll put together a PR.

**Comment by paulrosen** [comment]
For the modern syntax, right now the supported browsers are:
```
This works in any browser that supports AudioContext, AudioContext.resume, and Promises. That does NOT include IE, but this will work on any other "modern" browser that is at least the following version: Firefox 40, Safari 9.1, Edge 13, and Chrome 43.
```
Amazingly enough, I run into a wide variety of old browsers still. I'd like to not cut out a browser unless there is a compelling reason. That said, I'm planning on putting typescript in so it will become a moot point.

It looks like fetch just barely misses according to caniuse. But a polyfill is possible.

Just eyeballing your suggestion, it looks good. I'd love a pull request and as simple an html file with the problem as possible.

**Comment by foxbunny** [comment]
> Amazingly enough, I run into a wide variety of old browsers still. I'd like to not cut out a browser unless there is a compelling reason. That said, I'm planning on putting typescript in so it will become a moot point.

I didn't mean to modernize the code, that's what I was trying to get across. It was just me being lazy. :)

>  That said, I'm planning on putting typescript in so it will become a moot point.

Tbh, I'm not a big fan of TypeScript. The code it generates can sometimes be as much as 10% slower than the same code compiled by Babel. The compiler makes me sad.

> Just eyeballing your suggestion, it looks good. I'd love a pull request and as simple an html file with the problem as possible.

Sure thing.

**Comment by foxbunny** [comment]
I'm currently adding a HTML page/demo that also serves as a repro for this bug. Initially I had trouble reproducing the issue using the Cooley's and Persian tunes that are used in other examples. Then it occurred to me that maybe those tunes are too simple to repro this issue (my application is for Jazz music so tunes tend to be more complex). I changed one of the demo tunes to [this one](https://abcnotation.com/tunePage?a=trillian.mit.edu/~jc/music/abc/Scand/new/Polka_fr_Sodermanland_D-64-5/0000) and was successful in reproducing the issue. I'm going to further simplify the demo app to see what other factors affect this.

I'm also noticing that all demos are tied to the bundles in the `dist` directory. What is your recommended way to test the demo against the actual source? I'm thinking just generate a new bundles and reset `dist` prior to committing, but if you have a better method it would be great.

**Comment by foxbunny** [comment]
Alright, I've narrowed it down to two factors:

1. The number of different notes that need to be loaded simultaneously
2. Presence of multiple synth instances that are primed in parallel using the same instrument

Anyway, this is just a post-mortem, so I am confident that I'm not implementing a fix that isn't needed. :)

**Comment by paulrosen** [comment]
Fixed in beta.37
